few additional improvements to recurring plans

PetrDlouhy opened the pull request:
- fix DoesNotExist exception when recurring plan is not present
- add plan_autorenew_at() helper function to provide unified implementation
- add RecurringUserPlan.card_masked_number field

This requires PR #113 to be merged first to preserve order of migrations.

Alir3z4 commented on plans/models.py:
```diff
@@ -261,11 +261,20 @@ def get_plan_extended_until(self, plan, pricing):
+        """
+        Helper function which calculates when the plan autorenewal will occur
+        """
+        PLANS_AUTORENEW_BEFORE_DAYS = getattr(settings, 'PLANS_AUTORENEW_BEFORE_DAYS', 0)
```
I think in method scope, variables should not be all capital case ?

Alir3z4 commented on plans/models.py:
```diff
@@ -261,11 +261,20 @@ def get_plan_extended_until(self, plan, pricing):
         """
-        self.recurring.delete()
+        if hasattr(self, 'recurring'):
+            self.recurring.delete()
```
I think it would be better to add some logging here. 
Also a single comment would make the operation clear.

For instance why `recurring` object is being deleted and later another one is being created for.

The comment/documentation will make the code much easier to reason about.

Alir3z4 replied:
Also a unit test for this edge case would be a great addition.

Alir3z4 commented on plans/models.py:
```diff
@@ -261,11 +261,20 @@ def get_plan_extended_until(self, plan, pricing):
             return self.expire
         return self.get_plan_extended_from(plan) + timedelta(days=pricing.period)
 
+    def plan_autorenew_at(self):
```
This method would benefit from a simple unit test.

Alir3z4 commented:
Great stuff, just did a review and requested some changes.

Thanks for the great work on this.

Alir3z4 commented on demo/requirements.txt:
```diff
@@ -1,12 +1,14 @@
-django==2.2.13
```
hmm, why django has been removed ?

PetrDlouhy replied:
Because it collides with Django dependency defined in tox.ini

Alir3z4 replied:
Oh, it's the demo, never mind then, sorry my bad.

PetrDlouhy commented:
@Alir3z4 I have corrected the defects you mentioned and added the unit tests. I also added a few more things:
- fix of the tox settings
- few presets for my development environment
- fix for the `django_next_url_mixin` dependency version

Alir3z4 commented on setup.cfg:
```diff
@@ -1,4 +1,7 @@
 python_files=test*.py
 addopts=--tb=short
+[flake8]
+max-line-length = 180
```
I'd suggest 120 will good in case of using type hinting etc, but 180 would be very long and especially for code review that would require horizontal scrolling.

PetrDlouhy replied:
I can do that, but there are already many lines longer than that in the code. Too many long lines errors in the report was obscuring other more important errors, that's why I added this to the `setup.cfg`

But I can remove this commit from the PR entirely and solve all flake issues in separate PR and add it to the Travis testing, if you are interested.

Alir3z4 replied:
Hmm, that's unfortunate.
In that case, 180 is OK, the line count can be dealt with in a separate pull request, handling those long lines would pollute the pull request.

Let's carry on then.

Alir3z4 commented on tox.ini:
```diff
@@ -1,6 +1,17 @@
 [tox]
```
Additions to this are great, lovely <3

Alir3z4 commented:
All good, this looks good to me!

Let's get it in then.
Thanks for the great stuff.
